src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java

```diff
@@ -343,6 +347,16 @@ public void improveRecipe(IRecipeStorage recipe, int count, ICitizenData citizen
                     Blocks.AIR);
 
                 replaceRecipe(recipe.getToken(), IColonyManager.getInstance().getRecipeManager().checkOrAddRecipe(storage));
+
+                // Expected parameters for RECIPE_IMPROVED are Job, Result, Ingredient, Citizen
+                for(PlayerEntity player :colony.getMessagePlayerEntities())
```

Comment:
pretty sure we had a LanguageHandler function to send it to all players, should check that out

Reply:
I did, and it didn't work with the replacements correctly. I looked at the source for it in Structurize, and it looks overly complicated. We probably need to review a bunch of our LanguageHandler stuff, and look at where it makes sense to use MessageFormatter rather than String.format.

Reply:
There should be two versions, one uses string.format, second is the big thing for weird formatting used back in old minecolonies and left for compatibility reasons (tho ray is still using it and I still hadn't got how it works)

Reply:
also you definitely should reuse the message instead of creating it every time

Reply:
You should be able to give the Languagehandler version the TranslateableTextComponent

Reply:
Only one of the functions in LanguageHandler.java takes a player collection, sendPlayersMessage. That's the one that creates a chat component and didn't work correctly.  

sendMessageToPlayer sends to an individual player, by creating a stringTextComponent wrapped around a translateKeyWithFormat, which is a simple string.Format on the resulting key. 

Nothing in LanguageHandler takes a ITextComponent as it's only parameter. 

I think what I'm doing here is actually the most correct way to accomplish this (except the message reuse, which I'll fix).  If I use any of the current methods of LanguageHandler, I'm implicitly recreating the message every time anyway.

Reply:
To be clear: I believe we need to add this functionality (in several methods) to LanguageHandler, and I will start on that change, once I figure out appropriate naming that doesn't create a prototype that conflicts with what is already there.